packages/flutter_tools/lib/src/ios/fallback_discovery.dart

```diff
@@ -135,18 +135,19 @@ class FallbackDiscovery {
     @required int  hostVmservicePort,
     @required String packageName,
   }) async {
-    int hostPort;
+    ForwardedPort forwardedPort;
     Uri assumedWsUri;
     try {
-      hostPort = await _portForwarder.forward(
+      forwardedPort = await _portForwarder.forward(
         assumedDevicePort,
         hostPort: hostVmservicePort,
       );
-      assumedWsUri = Uri.parse('ws://localhost:$hostPort/ws');
+      assumedWsUri = Uri.parse('ws://localhost:${forwardedPort?.hostPort}/ws');
     } on Exception catch (err) {
       _logger.printTrace(err.toString());
       _logger.printTrace('Failed to connect directly, falling back to mDNS');
       _sendFailureEvent(err, assumedDevicePort);
+      await _portForwarder.unforward(forwardedPort);
```

Comment:
Most of this review is signature changes.  This is the real bug fix.

Reply:
You can call `_portForwarder.dispose()` here, and avoid changing the signature of `forward`. That is, let the `DevicePortForwarder` clean itself up rather than trying to do it piecemeal.

Reply:
I don't think we want the side-effect of all ports already forwarded in the injected `DevicePortForwarder` to be unforwarded just because this combination of device and package ID couldn't be forwarded.

Reply:
Ah, right. I missed that it is injected.